code/modules/mob/mob_say.dm

```diff
@@ -100,22 +100,47 @@
 	return LINGHIVE_NONE
 
 /**
-  * Get the mode of a message
+  * Extracts and cleans message of any extenstions at the begining of the message
+  * Inserts the info into the passed list, returns the cleaned message
   *
   * Result can be
+  * * SAY_MODE (Things like aliens, channels that aren't channels)
   * * MODE_WHISPER (Quiet speech)
   * * MODE_SING (Singing)
   * * MODE_HEADSET (Common radio channel)
-  * * A department radio (lots of values here)
+  * * RADIO_EXTENSION the extension we're using (lots of values here)
+  * * RADIO_KEY the radio key we're using, to make some things easier later (lots of values here)
+  * * LANGUAGE_EXTENSION the language we're trying to use (lots of values here)
   */
-/mob/proc/get_message_mode(message)
-	var/key = message[1]
-	if(key == "#")
-		return MODE_WHISPER
-	else if(key == "%")
-		return MODE_SING
-	else if(key == ";")
-		return MODE_HEADSET
-	else if((length(message) > (length(key) + 1)) && (key in GLOB.department_radio_prefixes))
-		var/key_symbol = lowertext(message[length(key) + 1])
-		return GLOB.department_radio_keys[key_symbol]
+#define MESSAGE_MODS_LENGTH 6 //The amount of items we are looking for in the message
+/mob/proc/get_message_mods(message, list/mods)

```

Comment:
instead of this iterative approach, consider a single pass using the fact you know the very restricted set of characters that are modifying the said string, and as soon as you hit a character not in that set then you are no longer caring about the contents.

ie in pseudocode
```
while (current_char in set_of_possible_modifiers)
    apply effect of current char
    continue to next char, or if at end of string then break
```

note that if you are just caring about having a distinct set of the modifiers, then you don't need to use the table to maintain state (mods) but instead could just have a list with the |= operator

Reply:
The table is needed for access later down the line. I considered using a while loop, but I decided against it to prevent someone spamming ########### or to keep a small fuckup on my part from freezing the server